File: plat/ts/entities/platform.ts
import Entity from "../entity.js";
import * as Draw from 'draw';
import CollisionType from "../collisiontype.js";

class Platform extends Entity {
  moving: boolean;
  dim: number;
  speed: number;
  start: number;
  end: number;
  oneShot: boolean;

  constructor(args: any) {
    super(args);
    this.moving = args.properties?.Moving ?? true;
    this.name = args.properties?.Name ?? '';
    this.dim = args.properties?.Direction == 'Horizontal' ? 0 : 1;
    this.speed = args.properties?.Speed ?? 1;
    this.oneShot = args.properties?.OneShot ?? false;

    const a = this.pos[this.dim];
    const b = a + (args.properties?.Distance ?? 100);
    this.speed *= a > b ? -1 : 1;
    this.start = Math.min(a, b);
    this.end = Math.max(a, b);
  }

  update(ticks: number, dt: number) {
    if (!this.moving) return;

    this.moveSolid(this.dim == 0 ? this.speed : 0, this.dim == 1 ? this.speed : 0);

    if (this.pos[this.dim] <= this.start || this.pos[this.dim] >= this.end) {
      this.speed *= -1;
      if (this.oneShot) {
        this.moving = false;
        return;
      }
    }
  }

  draw() {
    Draw.setColor(255, 255, 0, 255);
    Draw.rect(this.pos[0], this.pos[1], this.size[0], this.size[1], false);
  }

  trigger(other: Entity) {
    this.moving = !this.moving;
  }
}

export { Platform };

File: csgame/entities/Platform.cs
using Slate2D;

[Spawnable]
class Platform : Entity
{
    bool Moving;
    uint Dim;
    float Speed;
    int Start, End;
    bool OneShot;

    public Platform(LDTKEntity ent) : base(ent)
    {
        Layer = Layer.Background;



        Moving = ent.Properties.GetValueOrDefault("Moving", null)?.Bool ?? true;
        Name = ent.Properties.GetValueOrDefault("Name", null)?.Str ?? "";
        Dim = ent.Properties.GetValueOrDefault("Direction", null)?.Str == "Horizontal" ? 0u : 1u;
        Speed = ent.Properties.GetValueOrDefault("Speed", null)?.Num ?? 1;
        OneShot = ent.Properties.GetValueOrDefault("OneShot", null)?.Bool ?? false;

        int a = Dim == 0 ? Pos.X : Pos.Y;
        int b = a + (ent.Properties["Distance"]?.NumI ?? 100);
        Speed *= a > b ? -1 : 1;
        Start = Math.Min(a, b);
        End = Math.Max(a, b);
    }

    public override void Die() { }
    public override void Hurt(int number) { }

    public override void Update(uint ticks, float dt)
    {
        if (!Moving) return;

        float x = Dim == 0 ? Speed : 0;
        float y = Dim == 1 ? Speed : 0;
        MoveSolid(x, y);

        ref var pos = ref (Dim == 0 ? ref Pos.X : ref Pos.Y);
        if (pos <= Start || pos >= End)
        {
            Speed *= -1;
            if (OneShot) Moving = false;
        }
    }

    public override void Activate(Entity other)
    {
        Moving = !Moving;
    }

    public override void Draw()
    {
        DC.SetColor(251, 242, 54, 255);
        DC.Rect(Pos.X, Pos.Y, Size.W, Size.H, false);
    }
}